base/array.jl

```diff
@@ -2329,19 +2329,39 @@ function findall(A)
 end
 
 # Allocating result upfront is faster (possible only when collection can be iterated twice)
-function findall(A::AbstractArray{Bool})
-    n = count(A)
+function _findall(f::Function, A::AbstractArray{Bool})
+    n = count(f, A)
     I = Vector{eltype(keys(A))}(undef, n)
+    _findall(f, I, A)
+end
+
+function _findall(f::Function, I::Vector, A::AbstractArray{Bool})
     cnt = 1
-    for (i,a) in pairs(A)
-        if a
-            I[cnt] = i
-            cnt += 1
-        end
+    len = length(I)
+    for (k, v) in pairs(A)
+        cnt > len && break
+        I[cnt] = k
+        cnt += f(v)
+    end
+    I
```

Comment:
We could handle impure `f` here (e.g. calling `rand`) by checking the length again, and returning only the used part of the vector if `cnt < len`.

Reply:
I guess it's better to not have uninitialized indices. I've added a check that `cnt == len` and a `resize!` if not, for both methods.